apps/emqx_durable_storage/src/emqx_ds.erl

```diff
@@ -114,6 +123,16 @@
 
 -type get_iterator_result(Iterator) :: {ok, Iterator} | undefined.
 
+%% An opaque term identifying a generation.  Each implementation will possibly add
+%% information to this term to match its inner structure (e.g.: by embedding the shard id,
+%% in the case of `emqx_ds_replication_layer').
+-opaque generation_id() :: term().
```

Comment:
Generation is a concept from the storage layer that becomes RankY at the `emqx_ds` layer. It's best not to redefine this name.
Could this type simply instead be a pair of `{RankX, RankY}`? (I called it `stream_rank()` on my branch)
For the builtin backend, at least, `stream_rank()` is guaranteed to be unique.

Reply:
No problem in perhaps using the existing `emqx_ds:stream_rank()` type for it.

But then it might introduce the requirement on other implementation that values of this type should also be unique, besides the ordering properties, right?

Reply:
For example: let's say we have a single generation using LTS, with an empty trie.  After the first few messages are stored/iterated over, one or more streams already appear, and after it learns new structure, further streams appear, all in the same generation.

Reply:
> We need the value to directly map to an unique generation so that we may drop it later, so can we say stream_rank() is indeed an unique identifier for the generation? I thought that a generation could hold several streams inside it.

True. When I was writing the remark, I forgot to mention that `stream_rank()` is not a good _name_, but semantically it's the entity that most closely matches the definition of the unit that the garbage collector operates on.

Reply:
I renamed this type to `generation_rank()`... How about that?

Reply:
Yes, let's stick with it for now. Renaming types is possible without breaking BPAPI, so we don't have to commit to any particular name.